### PR TITLE
Setup runtime deployment

### PR DESCRIPTION
**Pull request by Pacman99**
Pre-submit checklist:
- Branch
    - [ ] Tests are provided (if possible)
    - [ ] [Test report is updated](https://github.com/input-output-hk/marlowe-cardano/blob/main/marlowe/test/test-report.md) (if relevant)
    - [ ] Commit sequence broadly makes sense
    - [ ] Key commits have useful messages
    - [ ] Relevant tickets are mentioned in commit messages
    - [ ] Formatting, materialized Nix files, PNG optimization, etc. are updated
- PR
    - [ ] Self-reviewed the diff
    - [ ] Useful pull request description
    - [ ] Reviewer requested

### CODE REVIEW

File: nix/nomadChart.nix [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+        # ----------
+        group.marlowe-runtime = let
+          # work-around: we need to get rid of vector first
+          node' = (cardano.nomadCharts.cardano-node (args // {jobname = "node";})).job.node.group.cardano;
+          group = l.removeAttrs node' ["task"];
+          node = group // {task.node = node'.task.node;};
+        in
+          merge
+          # task.vector ...
+          (vector.nomadTask.default {
```
Comment (jhbertra) [inline nix/nomadChart.nix:82]: can you leave a link to the definition of `vector` from `bitte-cells`?

File: nix/nomadChart.nix [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+            inherit namespace;
+            endpoints = [
+              # prometheus metrics for cardano-graphql
```
Comment (jhbertra) [inline nix/nomadChart.nix:85]: Is this a copy-paste error?

File: nix/nomadChart.nix [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+            extra = extraVector;
+          })
+          (
+            merge node
+              {
+                count = scaling;
+                # service = append [
+                #   (import ./srv-cardano-graphql.nix {inherit namespace healthChecks;})
+                # ];
+                network.port.http.to = "3100";
```
Comment (jhbertra) [inline nix/nomadChart.nix:97]: What is this for?

File: nix/nomadChart.nix [hunk 4]
```diff
@@ -0,0 +1,240 @@
...
+            ];
+            extra = extraVector;
+          })
+          (
+            merge node
+              {
+                count = scaling;
+                # service = append [
+                #   (import ./srv-cardano-graphql.nix {inherit namespace healthChecks;})
+                # ];
```
Comment (jhbertra) [inline nix/nomadChart.nix:96]:
```suggestion
```

File: nix/nomadChart.nix [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+                    check = [
+                      {
+                        type = "tcp";
+                        port = "http";
+                        interval = "10s";
+                        timeout = "2s";
+                      }
+                    ];
+                  }
+                ];*/
```
Comment (jhbertra) [inline nix/nomadChart.nix:118]:
```suggestion
```

File: nix/nomadChart.nix [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+                  chainseekd = {
+                    env = {
+                      HOST = "127.0.0.1";
+                      PORT = "8090";
+                      QUERY_PORT = "8091";
+                      JOB_PORT = "8092";
+                      CARDANO_NODE_SOCKET_PATH = "/alloc/tmp/node.socket"; # figure out how to pass this from the cardano group
+                      WORKLOAD_CACERT = "/secrets/tls/ca.pem";
+                      WORKLOAD_CLIENT_KEY = "/secrets/tls/key.pem";
+                      WORKLOAD_CLIENT_CERT = "/secrets/tls/cert.pem";
```
Comment (jhbertra) [inline nix/nomadChart.nix:133]: What's this for?

File: nix/nomadChart.nix [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+                    };
+
+                    template =
+                       _utils.nomadFragments.workload-identity-vault {inherit vaultPkiPath;}
+                      ++ _utils.nomadFragments.workload-identity-vault-consul {inherit consulRolePath;};
```
Comment (jhbertra) [inline nix/nomadChart.nix:138]: What's this for?

File: nix/nomadChart.nix [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+                    config.image = ociNamer oci-images.chainseekd;
+                    config.ports = ["http"];
```
Comment (jhbertra) [inline nix/nomadChart.nix:141]: Does this task need to expose an HTTP server?

File: nix/nomadChart.nix [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+                      PORT = 9080;
+                      QUERY_PORT = 9081;
+                      SYNC_PORT = 9082;
+                      CHAINSEEKD_HOST = "127.0.0.1";
+                      CHAINSEEKD_PORT = chainseekd.env.PORT;
+                      CHAINSEEKD_QUERY_PORT = chainseekd.env.QUERY_PORT;
+                    };
+
+                    config.image = ociNamer oci-images.marlowe-history;
+                    config.ports = ["http"];
```
Comment (jhbertra) [inline nix/nomadChart.nix:171]: Does this task need to expose HTTP? What about the ports that the service listens on? How will we connect to those?

File: nix/nomadChart.nix [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+                      change_mode = "noop";
+                      env = true;
+                      policies = ["marlowe-runtime"];
+                    };
+                  };
+                  marlowe-tx = {
+                    env = {
+                      HOST = "127.0.0.1";
+                      PORT = 11080;
+                      CHAINSEEKD_HOST = "127.0.0.1";
```
Comment (jhbertra) [inline nix/nomadChart.nix:213]: So, these all run on the same node? That will work for now, but that won't be how stuff will scale in the future. We can cross that bridge when we get to it I suppose.

**Review by jhbertra — COMMENTED** [review]
Just a few questions

File: flake.nix [hunk 11]
```diff
@@ -4,7 +4,7 @@
   inputs = {
     flake-utils.url = "github:numtide/flake-utils";
 
-    nixpkgs.follows = "haskell-nix/nixpkgs-2205";
+    nixpkgs.follows = "haskell-nix/nixpkgs-unstable";
```
Comment (jhbertra) [inline flake.nix:7]: Why this change?

File: deploy/oci-images.nix [hunk 12]
```diff
@@ -0,0 +1,37 @@
+{ inputs }:
+let
+  inherit (inputs) std self nixpkgs;
+  inherit (nixpkgs.lib) removePrefix;
+  inherit (self) operables;
+  inherit (self.sourceInfo) lastModifiedDate;
+
+  mkImage = name: std.lib.ops.mkStandardOCI {
+    name = "registry.ci.iog.io/dapps-world-${name}";
```
Comment (shlevy) [inline deploy/oci-images.nix:9]: Should we use the public registry?

**Review by shlevy — APPROVED** [review]

File: deploy/oci-images.nix (hunk 12 above)
Comment (Pacman99) [reply deploy/oci-images.nix:9]: Yeah we could switch to only using those. I originally kept the internal ones because they were easier to iterate over, but the entrypoints/images have mostly stabilized.

**Comment by shlevy** [comment]
flake evaluation is *extremely* slow. I don't know if this is a change from what was prior or if it's limited to the ops shell, but would be good to see some first-pass comparisons to ensure we're not crippling normal dev flow

**Comment by jhbertra** [comment]
This PR should get merged as soon as we are happy with it - it needs updating but I'd rather leave that for a follow-up PR and get this infrastructure merged into `main` first. Here are a few outstanding issues that need to be addressed:

1. Hydra seems to not be picking this up, we need to figure out why.
2. As Shea mentioned, flake evaluation is extremely slow. We need to identify why this is and fix the problem before we can accept the change in main.

**Comment by jhbertra** [comment]
The Hydra evaluation is failing with an error https://ci.zw3rk.com/jobset/input-output-hk-marlowe-cardano/pullrequest-383#tabs-errors

**Comment by jhbertra** [comment]
@Pacman99 @shlevy I added a new commit that removes `nixpkgs-unstable` from the inputs. We weren't using it anywhere as far as I could tell (we pulled `norouter` from it but that was not used) and updated the flake.lock. The slow flake evaluation issues seem to have been resolved and now I'm waiting to see if it builds on CI. Please let me know if this will cause any problems.
